### PR TITLE
Updated library exports

### PR DESCRIPTION
**Pull request by mshabunin**
replaces #8073
build with opencv/opencv_contrib#996

### This pullrequest changes

- use GenerateExportHeader
- fixed problems with exports and imports

```
build_world:Win64 MSVS2017 OpenCL=ON
#build_shared:Win64 OpenCL=OFF
allow_multiple_commits=1
```

### CODE REVIEW

**Comment by vpisarev** [comment]
@mshabunin, looks like the patch needs to be updated; sorry for not merging it earlier :(

**Comment by mshabunin** [comment]
@vpisarev , this patch is not compatible with #8461, because I've removed all `CV_EXPORTS` statements in the _ts_ module. First we need to choose either #8461 (build and install _ts_ module in separate package during OpenCV build) either #8468 (separately build standalone _ts_ library). After that I will update this patch.

File: cmake/OpenCVModule.cmake [hunk 1]
```diff
@@ -879,6 +878,8 @@ macro(_ocv_create_module)
     COMPILE_PDB_OUTPUT_DIRECTORY ${LIBRARY_OUTPUT_PATH}
     LIBRARY_OUTPUT_DIRECTORY ${LIBRARY_OUTPUT_PATH}
     RUNTIME_OUTPUT_DIRECTORY ${EXECUTABLE_OUTPUT_PATH}
+    DEFINE_SYMBOL CVAPI_EXPORTS
+    POSITION_INDEPENDENT_CODE TRUE
```
Comment (alalek) [inline cmake/OpenCVModule.cmake]: "PIC" is needed for Python/Java bindings only.
For using OpenCV with applications it is not required at all.

File: cmake/OpenCVModule.cmake [hunk 2]
```diff
@@ -925,7 +921,7 @@ macro(_ocv_create_module)
   if(OPENCV_MODULE_${the_module}_HEADERS AND ";${OPENCV_MODULES_PUBLIC};" MATCHES ";${the_module};")
     foreach(hdr ${OPENCV_MODULE_${the_module}_HEADERS})
       string(REGEX REPLACE "^.*opencv2/" "opencv2/" hdr2 "${hdr}")
-      if(NOT hdr2 MATCHES "private" AND hdr2 MATCHES "^(opencv2/?.*)/[^/]+.h(..)?$" )
+      if(NOT hdr2 MATCHES "opencv2/${the_module}/private.*" AND hdr2 MATCHES "^(opencv2/?.*)/[^/]+.h(..)?$" )
```
Comment (alalek) [inline cmake/OpenCVModule.cmake]: Need to revert this change.

File: modules/core/include/opencv2/core/cvdef.h [hunk 3]
```diff
@@ -248,22 +248,18 @@ Cv64suf;
...
-#if (defined _WIN32 || defined WINCE || defined __CYGWIN__) && defined CVAPI_EXPORTS
-#  define CV_EXPORTS __declspec(dllexport)
-#elif defined __GNUC__ && __GNUC__ >= 4
-#  define CV_EXPORTS __attribute__ ((visibility ("default")))
-#else
-#  define CV_EXPORTS
+#include "opencv2/cv_exports.h"
+
+// fix for Android FAT_JAVA_LIB: symbols should have default visibility even in static libraries
+#ifdef __ANDROID__
```
Comment (alalek) [inline modules/core/include/opencv2/core/cvdef.h]: Comment is not consistent with following code.
Android application native library should not export OpenCV API

File: modules/core/include/opencv2/core/cvdef.h [hunk 4]
```diff
@@ -248,22 +248,18 @@ Cv64suf;
...
-#    define CV_DEPRECATED __attribute__ ((deprecated))
-#  elif defined(_MSC_VER)
-#    define CV_DEPRECATED __declspec(deprecated)
-#  else
-#    define CV_DEPRECATED
-#  endif
+#ifdef _MSC_VER
+#define CV_EXPORTS_TEMPLATE
+#else
+#define CV_EXPORTS_TEMPLATE CV_EXPORTS
```
Comment (alalek) [inline modules/core/include/opencv2/core/cvdef.h]: `CV_EXPORTS_TEMPLATE` is not expected by bindings parser.

Comment (mshabunin) [reply modules/core/include/opencv2/core/cvdef.h]: Currently bindings generators do not support templates, so we don't need it for now. We can easily add `CV_EXPORTS_TEMPLATE_W` macro if such functionality will be implemented.

File: modules/core/include/opencv2/core/cvdef.h (hunk 3 above)
Comment (mshabunin) [reply modules/core/include/opencv2/core/cvdef.h]: We build native static libraries and "fat Java lib" for Android in one build (`-DBUILD_SHARED_LIBS=OFF -DBUILD_FAT_JAVA_LIB=ON`). We need to export all symbols for the second artifact. Currently all symbols are exported in static builds on all platforms, this PR changes this behavior for all platforms except Android. I can only suggest reworking "fat Java lib" artifact, so that it will be produced same way as _opencv_world_ module (`-DBUILD_SHARED_LIBS=ON -DBUILD_opencv_world=ON -DBUILD_FAT_JAVA_LIB=ON`), but it cat take more time to investigate.

File: cmake/OpenCVModule.cmake (hunk 1 above)
Comment (mshabunin) [reply cmake/OpenCVModule.cmake]: This is a port of block removed from _OpenCVCompilerOptions.cmake_. We always enable PIC because static libraries can be used to build some shared libraries, I wanted to preserve this ability.

File: modules/core/include/opencv2/core/cvdef.h (hunk 3 above)
Comment (alalek) [reply modules/core/include/opencv2/core/cvdef.h]: Fat Java library is artifact for Android manager which is deprecated.
If application doesn't use Java API (doesn't require libopencv_java.so), so there is no reason to bundle whole OpenCV code into Android application.
More accurate way is no link with application JNI library with static OpenCV libraries - libopencv_core.a/etc. So we need to drop global exported symbols from these OpenCV static libraries.

So, there are 2 different builds:
- static libopencv_<>.a with global exported symbols => libopencv_java.so
- static libopencv_<>.a without global exported symbols and without any libopencv_java.so. Useful for using OpenCV API from native C++ code only (without any OpenCV Java bindings).

**Comment by alalek** [comment]
I still do not understand what kind of problem should be resolved by this patch.
It removes OpenCV's CV_EXPORTS (and removes ability to adjust/customize this macro) and replaces with some CMake stuff (which probably depends on used CMake version). I believe this approach has very limited usage. Android "block" is an example that CMake can't target anything, especially cross-platform cases with custom toolchains.

**Comment by mshabunin** [comment]
This patch does not remove CV_EXPORTS macro, it changes the macro behavior depending on build configuration: dynamic builds will have same definitions as current version does, static builds will set it to empty. 
The _GenerateExportHeader_ module from cmake is rather simple and we can replace it with our own generated header when we need it, but I think that at this moment we can use it as-is. Problem with Android block takes its' roots in _opencv_java_ module and "non-trivial" "fat Java lib" feature implementation. We can remove it if we will rework the Android package.
